### PR TITLE
Longw/aks semver

### PR DESCRIPTION
**Pull request by wanlonghenry**
semver implementation for AKS

### CODE REVIEW

File: .pipelines/azure_pipeline_mergedbranches.yaml [hunk 1]
```diff
@@ -27,12 +27,11 @@ jobs:
     name: Azure-Pipelines-CI-Test-EO
   steps:
   - task: ComponentGovernanceComponentDetection@0
-
+  ## commit will be in format 3.0.0-14-g833be407
```
Comment (ganga1980) [inline .pipelines/azure_pipeline_mergedbranches.yaml:30]: where do you specify that commit id has to be in this format "3.0.0-14-g833be407"?

File: deployment/agent-deployment/ServiceGroupRoot/Parameters/ContainerInsights.Windows.Parameters.json [hunk 2]
```diff
@@ -40,8 +40,8 @@
               "value": "__AGENT_IMAGE_TAG_SUFFIX__"
             },
             {
-                "name": "AGENT_IMAGE_FULL_PATH",
-                "value": "public/azuremonitor/containerinsights/__AGENT_RELEASE__:win-__AGENT_RELEASE____AGENT_IMAGE_TAG_SUFFIX__"
+              "name": "AGENT_IMAGE_FULL_PATH",
+              "value": "public/azuremonitor/containerinsights/__AGENT_RELEASE__:win-__AGENT_IMAGE_TAG_SUFFIX__"
```
Comment (ganga1980) [inline deployment/agent-deployment/ServiceGroupRoot/Parameters/ContainerInsights.Windows.Parameters.json:44]: Lets confirm with Robbie about the contract we are going with for windows image and also preview images for the toggles.

File: .pipelines/azure_pipeline_mergedbranches.yaml (hunk 1 above)
Comment (wanlonghenry) [reply .pipelines/azure_pipeline_mergedbranches.yaml:30]: If the release tag we set are the semver format like x.x.x, then the   'git describe' command will return this format

File: .pipelines/release-agent.sh [hunk 3]
```diff
@@ -4,11 +4,12 @@
 
 # These are plain pipeline variable which can be modified anyone in the team
 # AGENT_RELEASE=cidev
-# AGENT_IMAGE_TAG_SUFFIX=07222021
+# AGENT_IMAGE_TAG_SUFFIX=3.0.0
 
 #Name of the ACR for ciprod & cidev images
 ACR_NAME=containerinsightsprod.azurecr.io
-AGENT_IMAGE_FULL_PATH=${ACR_NAME}/public/azuremonitor/containerinsights/${AGENT_RELEASE}:${AGENT_RELEASE}${AGENT_IMAGE_TAG_SUFFIX}
+## possible change
```
Comment (ganga1980) [inline .pipelines/release-agent.sh:11]: this comment can be removed.

File: .pipelines/azure_pipeline_mergedbranches.yaml (hunk 1 above)
Comment (ganga1980) [inline .pipelines/azure_pipeline_mergedbranches.yaml:30]: is this telemetry in AI matching with this image tag for both dev & prod images?

File: .pipelines/release-agent.sh (hunk 3 above)
Comment (wanlonghenry) [reply .pipelines/release-agent.sh:11]: removed

File: deployment/agent-deployment/ServiceGroupRoot/Scripts/pushAgentToAcr.sh [hunk 4]
```diff
@@ -22,10 +22,10 @@ fi
 
 if [[ "$AGENT_IMAGE_FULL_PATH" == *"win-"* ]]; then
   echo "checking windows tags"
-  TAG_EXISTS=$(echo $MCR_TAG_RESULT | jq '.tags | contains(["'"win-$AGENT_RELEASE$AGENT_IMAGE_TAG_SUFFIX"'"])')
+  TAG_EXISTS=$(echo $MCR_TAG_RESULT | jq '.tags | contains(["'"win-$AGENT_IMAGE_TAG_SUFFIX"'"])')
 else
   echo "checking linux tags"
-  TAG_EXISTS=$(echo $MCR_TAG_RESULT | jq '.tags | contains(["'":$AGENT_RELEASE$AGENT_IMAGE_TAG_SUFFIX"'"])')
+  TAG_EXISTS=$(echo $MCR_TAG_RESULT | jq '.tags | contains(["'":IMAGE_TAG_SUFFIX"'"])')
```
Comment (pfrcks) [inline deployment/agent-deployment/ServiceGroupRoot/Scripts/pushAgentToAcr.sh:28]: this seems to be incorrect. Should this be AGENT_IMAGE_TAG_SUFFIX?

File: .pipelines/release-agent.sh [hunk 5]
```diff
@@ -4,11 +4,11 @@
 
 # These are plain pipeline variable which can be modified anyone in the team
```
Comment (pfrcks) [inline .pipelines/release-agent.sh:5]: I don't think we are using this file anymore. Please confirm and if true remove this file

**Comment by pfrcks** [comment]
we no longer are using files under agent-deployment and multiarch-agent-deployment since Janvi created the mergedbranches one. please confirm this with Janvi and clean up the other two folders

File: .pipelines/azure_pipeline_mergedbranches.yaml [hunk 6]
```diff
@@ -27,12 +27,11 @@ jobs:
...
   steps:
   - task: ComponentGovernanceComponentDetection@0
-
+  ## commit will be in format Major.Minor.Patch-CommitAheadCount-CommitId
   - bash: |
-      commit=$(git rev-parse --short HEAD)
-      datetime=$(date +'%m%d%Y')
-      linuxImagetag="$datetime"-"$commit"
-      windowsImageTag=win-"$datetime"-"$commit"
+      commit=$(git describe)
```
Comment (pfrcks) [inline .pipelines/azure_pipeline_mergedbranches.yaml:32]: can you respond what the tags will look like for 

1. prod release
2. our dev builds
3. preview prod release

File: deployment/agent-deployment/ServiceGroupRoot/Scripts/pushAgentToAcr.sh (hunk 4 above)
Comment (wanlonghenry) [reply deployment/agent-deployment/ServiceGroupRoot/Scripts/pushAgentToAcr.sh:28]: changed

File: .pipelines/azure_pipeline_mergedbranches.yaml (hunk 6 above)
Comment (wanlonghenry) [reply .pipelines/azure_pipeline_mergedbranches.yaml:32]: added in comments above

File: .pipelines/release-agent.sh (hunk 5 above)
Comment (wanlonghenry) [reply .pipelines/release-agent.sh:5]: this is the script for manually push to mcr path for arc release, we might need it for reference

File: .pipelines/azure_pipeline_mergedbranches.yaml (hunk 1 above)
Comment (wanlonghenry) [reply .pipelines/azure_pipeline_mergedbranches.yaml]: the agent version will be in dev build format in AI telemetry like 3.1.0-12-commitId   while the agent version will be like 3.1.0

**Review by pfrcks — APPROVED** [review]

File: .pipelines/azure_pipeline_mergedbranches.yaml (hunk 1 above)
Comment (ganga1980) [reply .pipelines/azure_pipeline_mergedbranches.yaml]: what will be the version for the prod release in the telemetry?

File: .pipelines/release-agent.sh [hunk 7]
```diff
@@ -4,11 +4,11 @@
...
-# AGENT_IMAGE_TAG_SUFFIX=07222021
+# AGENT_IMAGE_TAG_SUFFIX=3.0.0
 
 #Name of the ACR for ciprod & cidev images
 ACR_NAME=containerinsightsprod.azurecr.io
-AGENT_IMAGE_FULL_PATH=${ACR_NAME}/public/azuremonitor/containerinsights/${AGENT_RELEASE}:${AGENT_RELEASE}${AGENT_IMAGE_TAG_SUFFIX}
+AGENT_IMAGE_FULL_PATH=${ACR_NAME}/public/azuremonitor/containerinsights/${AGENT_RELEASE}:${AGENT_IMAGE_TAG_SUFFIX}
 AGENT_IMAGE_TAR_FILE_NAME=agentimage.tar.gz
 
 if [ -z $AGENT_IMAGE_TAG_SUFFIX ]; then
```
Comment (ganga1980) [inline .pipelines/release-agent.sh:14]: we should move the error checks to the top i.e. before line 9.

File: README.md [hunk 8]
```diff
@@ -342,8 +342,8 @@ Integrated to Azure DevOps release pipeline for the ci_prod branch. With this, f
 
 When releasing the agent, we have a separate Azure DevOps pipeline which needs to be run to publish the image to prod MCR and our PROD AKS clusters. 
 
-For development, agent image will be in this format mcr.microsoft.com/azuremonitor/containerinsights/cidev:`<MM><DD><YYYY>`-<git-commit-id>.
-For releases, agent will be in this format mcr.microsoft.com/azuremonitor/containerinsights/ciprod:ciprod`<MM><DD><YYYY>-<git-commit-id>`.
+For development, agent image will be in this format mcr.microsoft.com/azuremonitor/containerinsights/cidev:`Major.Minor.Patch`-CommitAheadCount-<git-commit-id>.
+For releases, agent will be in this format mcr.microsoft.com/azuremonitor/containerinsights/ciprod:`Major.Minor.Patch`.
```
Comment (ganga1980) [inline README.md:346]: lets cover formats for both windows & Linux images.

File: ReleaseProcess.md [hunk 9]
```diff
@@ -2,7 +2,7 @@
 
 # 1. Production agent image and automatic deployment to CIPROD cluster
 
-Here are the high-level instructions to get the CIPROD`<MM><DD><YYYY>` image for the production release
+Here are the high-level instructions to get the CIPROD`MAJOR.MINOR.PATCH` image for the production release
```
Comment (ganga1980) [inline ReleaseProcess.md:5]: CIPROD prefix doesnt exist any more for the image tag.  so remove CIPROD from here.

File: .pipelines/azure_pipeline_mergedbranches.yaml (hunk 1 above)
Comment (wanlonghenry) [reply .pipelines/azure_pipeline_mergedbranches.yaml]: image

Comment (wanlonghenry) [reply .pipelines/azure_pipeline_mergedbranches.yaml]: addressed 
image

File: .pipelines/release-agent.sh (hunk 7 above)
Comment (wanlonghenry) [reply .pipelines/release-agent.sh:14]: moved

File: README.md (hunk 8 above)
Comment (wanlonghenry) [reply README.md:346]: added

File: ReleaseProcess.md (hunk 9 above)
Comment (wanlonghenry) [reply ReleaseProcess.md:5]: change to CIPROD:

File: kubernetes/windows/Dockerfile-dev-image [hunk 10]
```diff
@@ -34,6 +34,7 @@ COPY ./amalogswindows/installer/scripts/rubyKeepCertificateAlive/*.rb /etc/fluen
 COPY ./amalogswindows/ruby/ /etc/fluent/plugin/
 
 ENV AGENT_VERSION ${IMAGE_TAG}
+
```
Comment (pfrcks) [inline kubernetes/windows/Dockerfile-dev-image:37]: unrelated change? can be reverted

Comment (pfrcks) [inline kubernetes/windows/Dockerfile-dev-image:37]: unrelated change, can be reverted

File: .pipelines/azure_pipeline_mergedbranches.yaml [hunk 11]
```diff
@@ -27,15 +27,27 @@ jobs:
...
   - bash: |
-      commit=$(git rev-parse --short HEAD)
-      datetime=$(date +'%m%d%Y')
-      linuxImagetag="$datetime"-"$commit"
-      windowsImageTag=win-"$datetime"-"$commit"
+      commit=$(git describe) 
+      linuxImagetag="$commit"
+      windowsImageTag=win-"$commit"
+
+      tagConnector="-"
```
Comment (pfrcks) [inline .pipelines/azure_pipeline_mergedbranches.yaml:38]: not really clear here? it looks like telemetryTag is same as prefix? is there a scenario where it's not the same?

File: .pipelines/azure_pipeline_mergedbranches.yaml [hunk 12]
```diff
@@ -117,14 +130,14 @@ jobs:
         az acr login -n ${{ variables.containerRegistry }}
 
         if [ "$(Build.Reason)" != "PullRequest" ]; then
-          docker buildx build --platform linux/amd64,linux/arm64 --tag ${{ variables.repoImageName }}:$(linuxImagetag) -f kubernetes/linux/Dockerfile.multiarch --metadata-file $(Build.ArtifactStagingDirectory)/linux/metadata.json --build-arg IMAGE_TAG=$(linuxImagetag) --build-arg GOLANG_BASE_IMAGE=$(GOLANG_BASE_IMAGE) --build-arg CI_BASE_IMAGE=$(CI_BASE_IMAGE) --push .
+          docker buildx build --platform linux/amd64,linux/arm64 --tag ${{ variables.repoImageName }}:$(linuxImagetag) -f kubernetes/linux/Dockerfile.multiarch --metadata-file $(Build.ArtifactStagingDirectory)/linux/metadata.json --build-arg IMAGE_TAG=$(linuxTelemetryTag) --build-arg GOLANG_BASE_IMAGE=$(GOLANG_BASE_IMAGE) --build-arg CI_BASE_IMAGE=$(CI_BASE_IMAGE) --push .
 
           docker pull ${{ variables.repoImageName }}:$(linuxImagetag)
```
Comment (pfrcks) [inline .pipelines/azure_pipeline_mergedbranches.yaml:135]: the actual image tag is different from the telemetry tag? is this intentional? looks like imagetag will be something 3.1.0-3sdfd while telemetry tag is 3.1.0?

File: .pipelines/azure_pipeline_mergedbranches.yaml [hunk 13]
```diff
@@ -117,14 +130,14 @@ jobs:
...
         else
-          docker buildx build --platform linux/amd64,linux/arm64 --tag ${{ variables.repoImageName }}:$(linuxImagetag) -f kubernetes/linux/Dockerfile.multiarch --metadata-file $(Build.ArtifactStagingDirectory)/linux/metadata.json --build-arg IMAGE_TAG=$(linuxImagetag) --build-arg GOLANG_BASE_IMAGE=$(GOLANG_BASE_IMAGE) --build-arg CI_BASE_IMAGE=$(CI_BASE_IMAGE) .
+          docker buildx build --platform linux/amd64,linux/arm64 --tag ${{ variables.repoImageName }}:$(linuxImagetag) -f kubernetes/linux/Dockerfile.multiarch --metadata-file $(Build.ArtifactStagingDirectory)/linux/metadata.json --build-arg IMAGE_TAG=$(linuxTelemetryTag) --build-arg GOLANG_BASE_IMAGE=$(GOLANG_BASE_IMAGE) --build-arg CI_BASE_IMAGE=$(CI_BASE_IMAGE) .
```
Comment (pfrcks) [inline .pipelines/azure_pipeline_mergedbranches.yaml:137]: same as above

Comment (pfrcks) [reply .pipelines/azure_pipeline_mergedbranches.yaml:137]: also in the win stages

**Review by ganga1980 — DISMISSED** [review]

**Review by pfrcks — CHANGES_REQUESTED** [review]

File: .pipelines/azure_pipeline_mergedbranches.yaml (hunk 11 above)
Comment (pfrcks) [reply .pipelines/azure_pipeline_mergedbranches.yaml:38]: eg assuming commit is 3.1.1-22-g
linuxImageTag = 3.1.1-22-g

prefix = 3.1.1
index = 5
telemetryTag = 3.1.1

therefore, prefix is same as telemetryTag? then why the extra steps? am I missing something

Comment (wanlonghenry) [reply .pipelines/azure_pipeline_mergedbranches.yaml]: addressed https://github.com/microsoft/Docker-Provider/pull/876/commits/7265fe9609f5643347a19d0c2ee2802ba880269d

File: .pipelines/azure_pipeline_mergedbranches.yaml (hunk 12 above)
Comment (wanlonghenry) [reply .pipelines/azure_pipeline_mergedbranches.yaml:144]: addressed https://github.com/microsoft/Docker-Provider/pull/876/commits/7265fe9609f5643347a19d0c2ee2802ba880269d

File: .pipelines/azure_pipeline_mergedbranches.yaml (hunk 13 above)
Comment (wanlonghenry) [reply .pipelines/azure_pipeline_mergedbranches.yaml:146]: addressed in https://github.com/microsoft/Docker-Provider/pull/876/commits/7265fe9609f5643347a19d0c2ee2802ba880269d

File: kubernetes/windows/Dockerfile [hunk 14]
```diff
@@ -72,7 +70,10 @@ COPY ./amalogswindows/installer/scripts/rubyKeepCertificateAlive/*.rb /etc/fluen
 #Copy fluentd ruby plugins
 COPY ./amalogswindows/ruby/ /etc/fluent/plugin/
 
+# default value will be overwritten by pipeline 
+ARG IMAGE_TAG=win-ciprod01092023-864ecd78
```
Comment (ganga1980) [inline kubernetes/windows/Dockerfile:74]: merge latest changes from ci_prod

Comment (wanlonghenry) [reply kubernetes/windows/Dockerfile]: addressed

**Review by ganga1980 — APPROVED** [review]
